test/geocoders/here.py

```diff
@@ -0,0 +1,138 @@
+import unittest
+
+from geopy.compat import u
+from geopy.geocoders import Here
+from geopy.point import Point
+from test.geocoders.util import GeocoderTestBase, env
+
+
+class HereTestCaseUnitTest(GeocoderTestBase):
+
+    def test_user_agent_custom(self):
+        geocoder = Here(
+            app_id='DUMMYID1234',
+            app_code='DUMMYCODE1234',
+            user_agent='my_user_agent/1.0'
+        )
+        self.assertEqual(geocoder.headers['User-Agent'], 'my_user_agent/1.0')
+
+
+@unittest.skipUnless(
+    bool(env.get('HERE_APP_ID')),
+    "No HERE_APP_ID env variable set"
+)
+@unittest.skipUnless(
+    bool(env.get('HERE_APP_CODE')),
+    "No HERE_APP_CODE env variable set"
+)
+class HereTestCase(GeocoderTestBase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.geocoder = Here(
+            format_string='%s',
+            scheme='http',
+            app_id=env['HERE_APP_ID'],
+            app_code=env['HERE_APP_CODE'],
+            timeout=10
+        )
+
+    def test_geocode_empty_result(self):
+        """
+        Here.geocode empty results should be graciously handled.
+        """
+        self.geocode_run(
+            {"query": "xqj37"},
+            {},
+            expect_failure=True
+        )
+
+    def test_geocode(self):
+        """
+        Here.geocode
+        """
+        self.geocode_run(
+            {"query": "435 north michigan ave, chicago il 60611 usa"},
+            {"latitude": 41.890, "longitude": -87.624}
+        )
+
+    def test_geocode_structured(self):
+        """
+        Here.geocode
+        """
+        query = {
+            "street": "north michigan ave",
+            "housenumber": "435",
+            "city": "chicago",
+            "state": "il",
+            "postalcode": 60611,
+            "country": "usa"
+        }
+        self.geocode_run(
+            {"query": query},
+            {"latitude": 41.890, "longitude": -87.624}
+        )
+
+    def test_geocode_unicode_name(self):
+        """
+        Here.geocode unicode in Japanese for Paris. (POIs not included.)
+        """
+        self.geocode_run(
+            {"query": u("\u30d1\u30ea")},
+            {"latitude": 48.85718, "longitude": 2.34141}
+        )
+
+    def test_geocode_shapes(self):
+        """
+        Here.geocode using additional data parameter (postal code shapes)
+        """
+        address_string = "435 north michigan ave, chicago il 60611 usa"
+        res = self._make_request(
```

Comment:
Could you please remove `_make_request` usages?

```
res = self.geocode_run(
    {"query": address_string, "additional_data": "IncludeShapeLevel,postalCode"},
    {"latitude": 41.89035, "longitude": -87.62333},
)
shape_value = res.raw['Location']['Shape']['Value']
self.assertTrue(shape_value.startswith('MULTIPOLYGON ((('))
```

Reply:
Ok, sure. BTW, I wondered if it would be meaningful to add tests on the resulting shapes, but that would be easier with additional packages (like `shapely.wkt` or something smaller). What's the policy of adding other packages for testing only?

Reply:
I'm not sure that it's needed. Geopy doesn't work with shapes, so it's the API's responsibility to ensure that the returned shapes are valid. For geopy it should be enough to assume that if it *looks* like a shape (like you did with checking the prefix), then it's a valid shape.

As for the policy, the less deps the better :)

Reply:
Sounds good.